fms/modules/attention.py

```diff
@@ -267,6 +268,66 @@ def _sdpa_compute_op(
     return attn
 
 
+def _sdpa_with_sinks_op(
+    query: torch.Tensor,
+    key_cache: torch.Tensor,
+    value_cache: torch.Tensor,
+    nheads: int,
+    kvheads: int,
+    p_dropout: float,
+    scale_factor: Optional[float],
+    **attn_kwargs,
+) -> torch.Tensor:
+    queries = query.transpose(2, 1)
+
+    if key_cache.shape[1] != kvheads and key_cache.shape[2] == kvheads:
+        key_cache = key_cache.transpose(2, 1)
+        value_cache = value_cache.transpose(2, 1)
+
+    # Expand kv so black-box attn will work
+    expansion = nheads // kvheads
+    # k/v: b h l d
+    if expansion != 1:
+        keys_e = key_cache.unsqueeze(2).expand(-1, -1, expansion, -1, -1).flatten(1, 2)
+        values_e = (
+            value_cache.unsqueeze(2).expand(-1, -1, expansion, -1, -1).flatten(1, 2)
+        )
+    else:
+        keys_e = key_cache
+        values_e = value_cache
+
+    attn_sinks = attn_kwargs.get("sinks", None)
+    sliding_window = attn_kwargs.get("sliding_window", None)
+
+    # https://github.com/openai/gpt-oss/blob/main/gpt_oss/torch/model.py#L153
+    # from gpt-oss open ai implementation
+    n_tokens, n_heads, q_mult, d_head = queries.shape
+    S = attn_sinks.reshape(1, -1, 1, 1).expand(
+        queries.shape[0], -1, queries.shape[-2], -1
+    )
+    S = S.transpose(0, 1).transpose(1, 2)
+    mask = torch.triu(queries.new_full((n_tokens, n_tokens), -float("inf")), diagonal=1)
+    if sliding_window and sliding_window > 0:
+        mask += torch.tril(
+            mask.new_full((n_tokens, n_tokens), -float("inf")),
+            diagonal=-sliding_window,
+        )
+    QK = torch.einsum("qhmd,khmd->hmqk", queries, keys_e)
+    QK *= scale_factor
+    QK += mask[None, None, :, :]
+    QK = torch.cat([QK, S], dim=-1)
+    W = torch.softmax(QK, dim=-1)
```

Comment:
Two comments:
1. I'm a little surprised we're realizing the full `O(seq_len^2)` activations here, but this is apparently also what HF and the gpt-oss repo do. 
2. We should subtract the max QK scores before performing the softmax for numerical stability:
```py
QK = torch.cat([QK, S], dim=-1)
QK = QK - QK.max(dim=-1, keepdim=True).values
```
This is fairly standard and important, though apparently [only HF does this](https://github.com/huggingface/transformers/blob/f22cb1e8689905091acb07e238b670aadde1e4ee/src/transformers/models/gpt_oss/modular_gpt_oss.py?plain=1#L232) and not the original gpt-oss repo.